### PR TITLE
Support serving ES6 JS to clients

### PR DESCRIPTION
**Pull request by andrey-git**
Add an option to serve ES6 JS to clients
Backend PR:  home-assistant/home-assistant#10474

In addition to `hass_frontend` package with all the frontend content, generate `hass_frontend_es6` that contains untranspiled* files.

(Not really untranspiled, as react JS is still transpiled to regular JS, and babel es2016 preset is used)

@balloob I didn't check that **hassio-panel** produces valid result.

### CODE REVIEW

**Comment by tomitm** [comment]
You can use [babel-preset-env](https://github.com/babel/babel/tree/master/experimental/babel-preset-env) instead of es2016, which lets you specify a set of browsers to target (via browserslist).

Then you can offer either a "modern browsers" option, or default to modern and make legacy the option.

File: rollup.config-es6.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+const DEV = !!JSON.parse(process.env.BUILD_DEV || 'true');
+const DEMO = !!JSON.parse(process.env.BUILD_DEMO || 'false');
+
+const plugins = [
+  babel({
+    babelrc: false,
+    presets: [
+      [
+        'es2016',
```
Comment (balloob) [inline rollup.config-es6.js]: I wonder if we can just remove any preset completely?

If we want to use a preset, we should use browser-env.

File: rollup.config-es6.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      ]
+    ],
+    plugins: [
+      'external-helpers',
+      'transform-object-rest-spread',
```
Comment (balloob) [inline rollup.config-es6.js]: This one too is in the modern browsers.

File: rollup.config-es6.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+const plugins = [
+  babel({
+    babelrc: false,
+    presets: [
+      [
+        'es2016',
+      ]
+    ],
+    plugins: [
+      'external-helpers',
```
Comment (balloob) [inline rollup.config-es6.js]: If we don't transpile, we don't need this.

File: rollup.config-es6.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      'transform-object-rest-spread',
+      [
+        'transform-react-jsx',
```
Comment (balloob) [inline rollup.config-es6.js]: This is the only thing we need :-)

File: gulp/tasks/default.js [hunk 5]
```diff
@@ -4,6 +4,7 @@ const runSequence = require('run-sequence');
 gulp.task('default', () => {
   return runSequence.use(gulp)(
     'clean',
-    'build'
+    'build',
+    'build_es6'
```
Comment (balloob) [inline gulp/tasks/default.js]: I know it's kind of too late now, but I think "modern" would have been a better suffix that would stand time as I can see us moving what we transpile.

File: rollup.config-es6.js (hunk 1 above)
Comment (balloob) [reply rollup.config-es6.js]: Using `babel-preset-env` which takes in a browserlist query.

We can support latest 2 versions of Chrome, Firefox and Safari?

http://browserl.ist/?q=last+2+firefox+versions%2C+last+2+chrome+versions%2C+last+2+safari+versions

Comment (balloob) [reply rollup.config-es6.js]: https://github.com/babel/babel/tree/master/experimental/babel-preset-env

File: rollup.config-es6.js (hunk 3 above)
Comment (balloob) [reply rollup.config-es6.js]: We need this if we do browser-env.

File: gulp/common/transform.js [hunk 6]
```diff
@@ -1,27 +1,33 @@
...
-    .pipe(sourcesHtmlSplitter.split())
-    .pipe(gulpif(/[^app]\.js$/, babel({
+  return pump([
+    stream,
+    sourcesHtmlSplitter.split(),
+    gulpif(/[^app]\.js$/, babel({
       sourceType: 'script',
       presets: [
-        ['es2015', { modules: false }]
+        es6 ? ['es2016'] : ['es2015', { modules: false }]
```
Comment (balloob) [inline gulp/common/transform.js]: Should this be browser-env too?

File: gulp/tasks/gen-service-worker.js [hunk 7]
```diff
@@ -3,31 +3,32 @@ Generate a caching service worker for HA
...
-var uglifyJS = require('uglify-js');
-
-const config = require('../config');
+const gulp = require('gulp');
+const crypto = require('crypto');
+const file = require('gulp-file');
+const fs = require('fs');
+const path = require('path');
+const swPrecache = require('sw-precache');
+// var uglifyJS = require('uglify-js');
```
Comment (balloob) [inline gulp/tasks/gen-service-worker.js]: stale comment.

File: gulp/tasks/default.js (hunk 5 above)
Comment (andrey-git) [reply gulp/tasks/default.js]: Maybe change to es5 and un-suffixed?

File: rollup.config-es6.js (hunk 2 above)
Comment (andrey-git) [reply rollup.config-es6.js]: Browsers recognize it, but rollup doesn't.

File: rollup.config-es6.js (hunk 3 above)
Comment (andrey-git) [reply rollup.config-es6.js]: Without it a warning is issued. Lets keep it.

File: gulp/common/transform.js (hunk 6 above)
Comment (andrey-git) [reply gulp/common/transform.js]: Removed completely

File: gulp/tasks/gen-service-worker.js (hunk 7 above)
Comment (andrey-git) [reply gulp/tasks/gen-service-worker.js]: Removed

File: rollup.config-es6.js (hunk 1 above)
Comment (andrey-git) [reply rollup.config-es6.js]: Removed completely

File: gulp/tasks/default.js (hunk 5 above)
Comment (balloob) [reply gulp/tasks/default.js]: Oh I like that more. ES5 is a fixed goal.

**Review by balloob — APPROVED** [review]

**Comment by balloob** [comment]
@pvizeli reminder that I need to fix the Hass.io build before the next release.

**Comment by andrey-git** [comment]
Ouch, now if someone syncs FE they will get a broken experience until BE is submitted too.

**Comment by balloob** [comment]
Yeah, sorry about that 😞  thanks for the quick fix!

**Pull request merged**